Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -0,0 +1,1011 @@
+#endif    
+    std::map<std::string, int> matchSet;
+    std::map<std::set<int>, int> linkerMatchSet;
+    //std::map<std::vector<int>, int > attachMatch;
+    for (size_t m = 0; m < permutation.size(); ++m) {  // for each molecule
+      std::map<int, RGroupData>::const_iterator rg = matches[m][permutation[m]].rgroups.find(l);
+      if (rg != matches[m][permutation[m]].rgroups.end()) {
+#ifdef DEBUG        
+        std::cerr << " RGroup: " << rg->second.smiles;
+#endif        
+          matchSet[rg->second.smiles]++;
+#ifdef DEBUG                  
+          std::cerr << " score: " << matchSet[rg->second.smiles] << std::endl;
+#endif          
+          // XXX Use fragment counts to see if we are linking cycles?
+          if(rg->second.smiles.find(".") == std::string::npos &&
+             rg->second.attachments.size() > 1) {
+            linkerMatchSet[rg->second.attachments]++;
+#ifdef DEBUG                              
+            std::cerr << " Linker Score: " << linkerMatchSet[rg->second.attachments]++ << std::endl;
+#endif            
+          }
+      }
+    }
+    // scoring system is the max # of matches per label
+    //  I.e. if we have 5 matches and 3 non-matches, the score for the
+    //  label is 5./8.
+    double tempScore = 1.;
+    int maxMatches = 0;
+    std::string m;
+    for (std::map<std::string, int>::const_iterator it = matchSet.begin();
+         it != matchSet.end(); ++it) {
+      if (it->second > maxMatches) {
+        tempScore *= it->second/(double)matches.size();
+        maxMatches = it->second;
+        m = it->first;
+      }
+    }
+#ifdef DEBUG                                  
+    std::cerr << "Max Matches :" << maxMatches << " at rgroup " << m << std::endl;
+#endif    
+    // overweight linkers with the same attachments points....
+    int maxLinkerMatches = 0;
+    for (std::map<std::set<int>, int>::const_iterator it = linkerMatchSet.begin();
+         it != linkerMatchSet.end(); ++it) {
+      if (it->second > 1) {
+        if (it->second > maxLinkerMatches)
+          maxLinkerMatches = it->second;
+      }
+    }
+#ifdef DEBUG    
+    std::cerr << "Max Linker Matches :" << maxLinkerMatches << std::endl;
+#endif
+    double increment = 1.0; // no change in score
+    double linkerIncrement = 1.0; // no change in score
+    if(maxLinkerMatches) {
+      linkerIncrement = (double)(maxLinkerMatches) / (double)matches.size();
+    } else if(!maxMatches) { // nothing happens
+    } else {
+      increment = tempScore;
+    }
+
+    score *= increment * linkerIncrement;
+#ifdef DEBUG    
+    std::cerr << "Increment: " << increment << " Linker_Increment: " << linkerIncrement << std::endl;
+    std::cerr << "increment*linkerIncrement: " << increment * linkerIncrement << std::endl;
+    std::cerr << "Score = " << score << std::endl;
+#endif    
+  }
+  
+  return score;
+}
+
+
+}
+
+
+struct RGroupDecompData {
+  // matches[mol_idx] == vector of potential matches
+  std::map<int, RWMol> cores;
```

Comment:
Maps are a bit special

```
Section 23.1.2#8 (associative container requirements):

"The insert members shall not affect the validity of iterators and references to the container, and the erase members shall invalidate only iterators and references to the erased elements."
```

I.e. their memory locations are stable.  The overhead here is a copy construct and I can live with that.  If it was a vector, there would be many arbitrary copies made, maps are stable in this regard.